java-tests/validating/src/test/kotlin/io/spine/test/options/setonce/SetOnceErrorMessageITest.kt

```diff
@@ -81,34 +79,42 @@ private fun <T : Any> assertCustom(fieldName: String, fieldType: String, value1:
  * so we also have to take this into account during assertions because
  * in `ConstraintViolation` they still arrive as Java enum constants.
  *
+ * @param template The error message template to check.
  * @param fieldName The field to set.
  * @param value1 The first field value to set.
  * @param value2 The second field value to set for triggering the exception.
- * @param expectedParams The list of params to check upon `ConstraintViolation.param`.
- * @param expectedFormat The format string to check upon `ConstraintViolation.msg_format`.
  */
 private fun <T : Any> Builder.assertErrorMessage(
+    template: (Int) -> String,
     fieldName: String,
+    fieldType: String,
     value1: T,
     value2: T,
-    expectedParams: List<String>,
-    expectedFormat: (Int) -> String
 ) {
     check(value1 != value2)
 
     val field = descriptorForType.field(fieldName)!!
+    val parentType = descriptorForType.fullName
     val exception = assertThrows<ValidationException> {
         setField(field, value1)
         setField(field, value2)
     }
 
     val violations = exception.constraintViolations.also { it.size shouldBe 1 }
     val violation = violations.first()
+
     with(violation) {
-        msgFormat shouldBe expectedFormat(field.index + 1)
-        paramList shouldBe expectedParams
+        message.withPlaceholders shouldBe template(field.index + 1)
+        message.placeholderValueMap shouldContainExactly mapOf(
+            "field.name" to fieldName,
```

Comment:
As discussed vocally, let's have something less prone to typos for operating the placeholder names than just string literals. E.g. constants, and probably some API on top of them, such as extension functions for `MessageTemplate.Builder` (for Kotlin API, at least).